### PR TITLE
Week3 step4

### PR DESCRIPTION
**Pull request by Xin128**
Finished Step 4:
Finished features:
1) get the input message from user
2) fetch it to the container on my web page

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 1]
```diff
@@ -31,12 +30,32 @@
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
-  ArrayList<String> msglist = new ArrayList<String> (Arrays.asList("Hi!","Ni Hao!","Bonjour!","Hola!"));
-  
+  ArrayList<String> msglist = new ArrayList<String>();
+  String jsonMsg;
```
Comment (warfa-google) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: add new line

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 2]
```diff
@@ -31,12 +30,32 @@
...
+    String inputMsg = getParameter(request, "comment-input", "");
+    msglist.add(inputMsg);
+    response.sendRedirect("/index.html");
+  }
+
+  /**
+   * @return the request parameter, or the default value if the parameter
+   *         was not specified by the client
+   */
+  private String getParameter(HttpServletRequest request, String name, String defaultValue) {
```
Comment (warfa-google) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: you can replace this by using the ternary operator. 
return request.parameter(name) == null ? defaultValue : value;

File: portfolio/src/main/webapp/script.js [hunk 3]
```diff
@@ -37,3 +37,10 @@ function getHellofromFetchedPractice() {
     document.getElementById('fetch-container').innerText = quote;
   });
 }
+
```
Comment (warfa-google) [inline portfolio/src/main/webapp/script.js:39]: make sure lines between function are consistent

File: portfolio/src/main/webapp/script.js [hunk 4]
```diff
@@ -37,3 +37,10 @@ function getHellofromFetchedPractice() {
...
+
+function getCommentInForm() {
```
Comment (warfa-google) [inline portfolio/src/main/webapp/script.js:44]: Also add a comment explaining what this function does

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 1 above)
Comment (warfa-google) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: what is this used for?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 5]
```diff
@@ -31,12 +30,32 @@
...
 
-  ArrayList<String> msglist = new ArrayList<String> (Arrays.asList("Hi!","Ni Hao!","Bonjour!","Hola!"));
-  
+  ArrayList<String> msglist = new ArrayList<String>();
+  String jsonMsg;
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
     response.setContentType("application/json;");
-    response.getWriter().println(new Gson().toJson(msglist));
+    response.getWriter().println(new Gson().toJson(msglist)) ;
```
Comment (warfa-google) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: space before semicolon

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 6]
```diff
@@ -31,12 +30,32 @@
...
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Get the input from the form.
+    String inputMsg = getParameter(request, "comment-input", "");
+    msglist.add(inputMsg);
```
Comment (warfa-google) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: do you want to add inputMsg to msgList if it's blank?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 7]
```diff
@@ -31,12 +30,32 @@
...
+    response.sendRedirect("/index.html");
+  }
+
+  /**
+   * @return the request parameter, or the default value if the parameter
+   *         was not specified by the client
```
Comment (cisxingg) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Extra spaces in between * and comment.

File: portfolio/src/main/webapp/script.js [hunk 8]
```diff
@@ -37,3 +37,10 @@ function getHellofromFetchedPractice() {
...
+  fetch('/data').then(response => response.text()).then((quote) => {
```
Comment (cisxingg) [inline portfolio/src/main/webapp/script.js]: nit: it is not returning a quote but instead a comment now, could update the variable name to make it clearer :)

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 2 above)
Comment (cisxingg) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Agreed with using ternary operator, but can probably keep:
String value = request.getParameter(name); and use value for the return statement.

Comment (cisxingg) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: But what could be better is to move the logic of checking null above, as Warfa mentioned, if null, we may not want to add it to the list.

**Review by cisxingg — COMMENTED** [review]
Looks pretty good!

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 2 above)
Comment (Xin128) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: To make better logic, I deleted the getParameter() method and directly check whether the string is empty to put it inside the arraylist

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 9]
```diff
@@ -31,12 +30,22 @@
...
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
     response.setContentType("application/json;");
     response.getWriter().println(new Gson().toJson(msglist));
   }
 
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Get the input from the form.
+    String inputMsg = request.getParameter("comment-input");
+    if (!inputMsg.isEmpty()) {
```
Comment (cisxingg) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:45]: Seems like getParameter() actually returns a null. Checking isEmpty() is not really the same as checking not null.

**Review by cisxingg — APPROVED** [review]
One comment, please resolve before merging :)

**Review by warfa-google — APPROVED** [review]
Fix CIS comment about isEmpty. Otherwise LGTM

**Pull request merged**